### PR TITLE
Cable::Server#active_connections_for and Cable::Server#subscribed_channels_for public checkup methods

### PR DESCRIPTION
**Pull request by rmarronnier**
From https://github.com/cable-cr/cable/issues/85, this is a draft proposal PR investigating the best way for the backend to check what connection / channel are active for a given token (ie user).

The maze of different `identifiers` must make me oblivious to a simpler (ie more efficient than a rough `#select == token`) way to get the info we need.

I've added some DRY suggestions.

Right now, this is a draft, it lacks proper specs.

I'm working to add them but if you have DX / code design remarks / etc. please tell, I'm really new to this codebase (and websockets in general) !

### CODE REVIEW

File: src/cable/connection.cr [hunk 1]
```diff
@@ -2,7 +2,7 @@ require "uuid"
 
 module Cable
   abstract class Connection
-    class UnathorizedConnectionException < Exception; end
+    class UnauthorizedConnectionException < Exception; end
```
Comment (jwoertink) [inline src/cable/connection.cr:5]: oops! :joy: good catch

**Comment by rmarronnier** [comment]
This PR is ready for review, but :

- I am not 100% convinced Cable need the two new public methods `active_connections_for` and `subscribed_channels_for`
They don't exist in the ruby version (by the way, does `cable-cr` aim to mirror the action cable backend api ?),  they might be misleading / misused when several Cable.server instances are spawn (most serious production cases ?).

That being said, writing these methods and the corresponding specs made me fix a bug (calling `connection.close` on a connection without channel subscriptions wasn't closing it), refactor some code and complete the server specs.

So, I won't fight for these methods as new public api if it "stalls" this PR as the rest of the code changes suggestions is more valuable IMO :-)

File: src/cable/connection.cr [hunk 2]
```diff
@@ -52,36 +52,43 @@ module Cable
...
+      Connection::CHANNELS.[connection_identifier].values
+    end
+
     def closed? : Bool
       socket.closed?
     end
 
     def close
-      return true unless Connection::CHANNELS.has_key?(connection_identifier)
+      if Connection::CHANNELS.has_key?(connection_identifier)
```
Comment (rmarronnier) [inline src/cable/connection.cr:65]: We shouldn't return before closing the connection socket in the case of a connection without channel subscriptions.

File: src/cable/connection.cr [hunk 3]
```diff
@@ -52,36 +52,43 @@ module Cable
       @connection_rejected = true
     end
 
+    def channels : Array(Channel)
```
Comment (jwoertink) [inline src/cable/connection.cr:55]: To ensure there's no ambiguity, should we qualify this `Channel` with the namespace `Cable::Channel` so it's not confused with Crystal's `Channel`?

**Review by jwoertink — COMMENTED** [review]
Looks like there's a bit of refactoring going on in here, but it all looks good from what I see :+1:  I left a comment, but otherwise I think this is good.

File: src/cable/connection.cr (hunk 3 above)
Comment (rmarronnier) [reply src/cable/connection.cr:55]: Good call, `Cable::Channel `is used in the codebase and is more explicit.

**Comment by rmarronnier** [comment]
> Looks like there's a bit of refactoring going on in here, but it all looks good from what I see 👍 I left a comment, but otherwise I think this is good.

Thanks for the review :-)

**Comment by jwoertink** [comment]
:man_facepalming:  Ameba!!!! lol

**Review by jwoertink — APPROVED** [review]

**Pull request merged**